pkg/envoy/registry/services.go

```diff
@@ -206,16 +225,17 @@ func (k *AsyncKubeProxyServiceMapper) handleServiceDelete(svc *v1.Service) {
 		Name:      svc.Name,
 		Namespace: svc.Namespace,
 	}
-	for cn, svcs := range k.servicesForCN {
+	for cn := range k.cnsForService[deleted] {
 		var rem []service.MeshService
-		for _, s := range svcs {
+		for _, s := range k.servicesForCN[cn] {
```

Comment:
^ My only beef with this would be, if `servicesForCN` was not a `map[certificate.CommonName][]service.MeshService` and instead was a `map[certificate.CommonName]map[service.MeshService]` you would simplify this code by just doing
`delete(k.servicesForCN[cn], deleted)`, and not rewriting the slice every time

if things can be uniquely indexed, I'd normally say to just go for it. It simplifies a lot of walks and linear searches in code later on (not even arguing performance, even code gets shorter and cleaner)

Reply:
I agree that would be easier to work with here. I kept the value as a slice here just so when it's queried we can return it as-is without doing copying it into a slice every time to preserve the existing API. I'll think I'll keep it this way for now and as a follow-up see if it's not a huge pain for `ListProxyServices()` to return a map vs. a slice.